draft/XXXX-multipart-upload/README.md

```diff
@@ -0,0 +1,115 @@
+following endpoints:
+
+- `POST /api/uploads` - creates a new upload object by selecting a random,
+non-preexisting SHA string, then creating an Arango collection keyed by the
+string. This endpoint returns the ID string.
+- `POST /api/uploads/<upload-id>/chunk?sequence=<sequence-number>` - uploads one
+chunk of data to the named upload object. This endpoint returns the sequence
+number, and may raise various status codes to indicate problems.
+- `POST /api/uploads/<upload-id>/finalize` - finalizes the IDed upload by
+preventing further calls to `POST /api/uploads/<upload-id>/chunk` and performing
+a checksum on the settled data.
+- `DELETE /api/uploads/<upload-id>` - removes the database collection associated
+with the given ID.
+
+### File-like interface to upload collection
+
+A Python class named `UploadReader` which implements the file-like interface
+for reading data from an upload collection will be used by the serverside to
+interact with the data uploaded to the database using the `/api/uploads`
+endpoints. It should be constructed with a string ID referring to an upload, and
+it should implement the `seek()`, `read()`, and `readline()` methods as though
+the various documents in the collection constitute a regular file as opened by
+Python's `open()` function.
+
+### Usage
+
+Currently, file uploads are done by file type using the `uploaders` API section.
+There, a file of the specified type is sent to a specific `POST
+/api/uploaders/<file-type>` endpoint, which in turn handles the full file data
+including validation and ingestion to an Arango database and collection.
+
+Under this RFC, file upload becomes detached from file ingestion. Now, the
+correct sequence of `/api/uploads` calls must be made in order to settle the
+data in the database in an intermediate form.
+
+Next, the client can invoke one of the uploader paths, which will in turn
+instantiate an `UploadReader` object. This file-like object can be used to
+perform validation and bulk ingestion into a Multinet data table proper. The
+main advantage of this usage is that an `UploadReader` does not exhaust its data
+iterator after one pass; it can be rewound by using the `.seek()` method. This
+means that both validation and ingestion can be performed without a memory
+blowup caused by exhausting an HTTP request body iterator into a string in
+memory.
+
+Finally, the client (or the uploader methods) can invoke the `DELETE` endpoint
+to remove the intermediate upload data, leaving the server in a clean state.
+
+## Questions
+
+1. How do we detect upload interruptions? What is the right way to resume such
+   an interrupted upload?
+2. What do we do with "stale" upload objects? One answer is to have a `vaccum`
+   endpoint that looks for upload objects older than a specific cutoff age and
+   remove them, then run that endpoint periodically. Better still, it could be
+   installed as a cronjob on the server directly.
```

Comment:
A cronjob seems appropriate

Reply:
We'll want to figure out if there's an easy way to incorporate that into provisioning. I believe it is pretty simple--in particular, that Ansible includes modules for doing this. And then we should discuss whether cronjobs are the best way to handle this.

Reply:
I think ideally we'd have a hands-on system in place for dealing with stale uploads, instead of using a cron job, as it introduces a devops burden, as well as a disconnect between our application and the database (i.e. the database is modifying things behind the server's back). Plus, it makes the application less portable, as if you were using some managed database for example, you wouldn't be able to just throw a cron job on there to deal with this. However, I'm not sure what we can really do at this point besides that, without adding more complexity anyway (like celery, etc.).

I've done cron job provisioning through Ansible before, and it isn't difficult, but does introduce some very deployment-specific devops details.

Reply:
I actually agree with your concerns. However, we're already chest-deep in devops-specificity, and we need some way to solve this problem. I'm totally open to ideas.

When you say "hands on", do you mean that we should execute a script manually every so often to clean up after stale uploads? If we have such a thing, why wouldn't we just automate that via a devops-deployed cronjob?

Reply:
Sorry, by "hands on", I meant that the server application is aware/in control of the methods used to clean up stale uploads.

Reply:
> Is the database hosted on heroku? If so, maybe this could be another option next to Ansible
> https://devcenter.heroku.com/articles/scheduler
> 
> That does make things dependent on Heroku as a host though so maybe it's not worth it. Still, thought it was at least worth mentioning.

Unfortunately, no. We're hosting that on AWS. Heroku doesn't have an offering for ArangoDB, I assume because the ArangoDB people themselves sell hosted services. Unfortunately (again), the ArangoDB hosting services offer a minimum service level that's much too large/expensive for our purposes.

Reply:
> Sorry, by "hands on", I meant that the server application is aware/in control of the methods used to clean up stale uploads.

How do we accomplish this? This sounds like an alternative plan to using cronjob via devops deployment of the DB.